pkg/apis/serving/v1/serviceorchestrator_types.go

```diff
@@ -0,0 +1,176 @@
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +genreconciler
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ServiceOrchestrator represents the orchestrator to launch the new revision and direct the traffic
+// in an incremental way.
+type ServiceOrchestrator struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec ServiceOrchestratorSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status ServiceOrchestratorStatus `json:"status,omitempty"`
+}
+
+// Verify that ServiceOrchestrator adheres to the appropriate interfaces.
+var (
+	// Check that we can create OwnerReferences to a ServiceOrchestrator.
+	_ kmeta.OwnerRefable = (*ServiceOrchestrator)(nil)
+
+	// Check that the type conforms to the duck Knative Resource shape.
+	_ duckv1.KRShaped = (*ServiceOrchestrator)(nil)
+)
+
+// TargetRevision holds the information of the revision for the current stage.
+type TargetRevision struct {
+	// RevisionName indicates RevisionName.
+	// +optional
+	RevisionName string `json:"revisionName,omitempty"`
+
+	// Direction indicates up or down.
+	// +optional
+	Direction string `json:"direction,omitempty"`
+
+	// TargetReplicas indicates an estimated number of replicas.
+	// +optional
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// IsLatestRevision indicates whether it is the last revision or not.
+	// +optional
+	IsLatestRevision *bool `json:"IslatestRevision,omitempty"`
+
+	// Percent indicates that percentage based routing should be used and
+	// the value indicates the percent of traffic that is routed to this
+	// Revision. `0` (zero) mean no traffic, `100` means all
+	// traffic.
+	// When percentage based routing is being used the follow rules apply:
+	// - the sum of all percent values must equal 100
+	// - when not specified, the implied value for `percent` is zero for
+	//   that particular Revision
+	// +optional
+	Percent *int64 `json:"percent,omitempty"`
+
+	// MinScale sets the lower bound for the number of the replicas.
+	// +optional
+	MinScale *int32 `json:"minScale,omitempty"`
+
+	// MaxScale sets the upper bound for the number of the replicas.
+	// +optional
+	MaxScale *int32 `json:"maxScale,omitempty"`
+}
+
+// StageTarget holds the information of all revisions during the transition for the current stage.
+type StageTarget struct {
+	// StageTargetRevisions holds the configured traffic distribution for the current stage.
+	// These entries will always contain RevisionName references.
+	// +optional
+	StageTargetRevisions []TargetRevision `json:"stageTargetRevisions,omitempty"`
+
+	// TargetFinishTime indicates target time to complete this target.
+	// +optional
+	TargetFinishTime apis.VolatileTime `json:"targetFinishTime,omitempty"`
+}
+
+// ServiceOrchestratorSpec holds the desired state of the ServiceOrchestrator (from the client).
+type ServiceOrchestratorSpec struct {
+	StageTarget `json:",inline"`
```

Comment:
stages in transition indicates that these should be status fields?

Reply:
There are fields service orchestrator will set for each stage. They are specs.